From dcd913e6e41f5ebdf3a5eb38d3dab65fc2f503a0 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:18:54 -0400
Subject: [net] netfilter: nf_tables: decrement chain use counter when replacing rules

Message-id: <20141021161854.22888.51472.stgit@localhost.localdomain>
Patchwork-id: 97967
O-Subject: [RHEL7 net PATCH 070/124] netfilter: nf_tables: decrement chain use counter when replacing rules
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: ac34b861979ec5057d686c890b1b8f8661e9b99f

 commit ac34b861979ec5057d686c890b1b8f8661e9b99f
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Tue Jun 10 10:53:02 2014 +0200

    netfilter: nf_tables: decrement chain use counter when replacing rules

    Thus, the chain use counter remains with the same value after the
    rule replacement.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 5586426..19f438d 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -1799,6 +1799,7 @@ static int nf_tables_newrule(struct sock *nlsk, struct sk_buff *skb,
     goto err2;
    }
    nft_rule_disactivate_next(net, old_rule);
+   chain->use--;
    list_add_tail_rcu(&rule->list, &old_rule->list);
   } else {
    err = -ENOENT;
@@ -1829,6 +1830,7 @@ err3:
   list_del_rcu(&nft_trans_rule(trans)->list);
   nft_rule_clear(net, nft_trans_rule(trans));
   nft_trans_destroy(trans);
+  chain->use++;
  }
 err2:
  nf_tables_rule_destroy(&ctx, rule);
-- 
1.7.1